Code/Framework/AzCore/AzCore/DOM/DomPatch.cpp

```diff
@@ -0,0 +1,968 @@
+            {
+                // Add -> Replace (if value already existed in an object) otherwise
+                // Add -> Remove
+                if (m_domPath.Size() > 0 && m_domPath[m_domPath.Size() - 1].IsKey())
+                {
+                    const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                    if (existingValue != nullptr)
+                    {
+                        return AZ::Success(PatchOperation::ReplaceOperation(m_domPath, *existingValue));
+                    }
+                }
+                return AZ::Success(PatchOperation::RemoveOperation(m_domPath));
+            }
+        case Type::Remove:
+            {
+                // Remove -> Add
+                const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                if (existingValue == nullptr)
+                {
+                    AZStd::string errorMessage = "Unable to invert DOM remove patch, source path not found: ";
+                    m_domPath.AppendToString(errorMessage);
+                    return AZ::Failure(AZStd::move(errorMessage));
+                }
+                return AZ::Success(PatchOperation::AddOperation(m_domPath, *existingValue));
+            }
+        case Type::Replace:
+            {
+                // Replace -> Replace (with old value)
+                const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                if (existingValue == nullptr)
+                {
+                    AZStd::string errorMessage = "Unable to invert DOM replace patch, source path not found: ";
+                    m_domPath.AppendToString(errorMessage);
+                    return AZ::Failure(AZStd::move(errorMessage));
+                }
+                return AZ::Success(PatchOperation::ReplaceOperation(m_domPath, *existingValue));
+            }
+        case Type::Copy:
+            {
+                // Copy -> Replace (with old value)
+                const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                if (existingValue == nullptr)
+                {
+                    AZStd::string errorMessage = "Unable to invert DOM copy patch, source path not found: ";
+                    m_domPath.AppendToString(errorMessage);
+                    return AZ::Failure(AZStd::move(errorMessage));
+                }
+                return AZ::Success(PatchOperation::ReplaceOperation(m_domPath, *existingValue));
+            }
+        case Type::Move:
+            {
+                // Move -> Replace, using the common ancestor of the two paths as the replacement
+                // This is not a minimal inverse, which would be two replace operations at each path
+                const Path& destPath = m_domPath;
+                const Path& sourcePath = GetSourcePath();
+
+                Path commonAncestor;
+                for (size_t i = 0; i < destPath.Size() && i < sourcePath.Size(); ++i)
```

Comment:
I think this can have potential performance implications down the road. If you consider a scenario where you are moving a simple value from outside of a big dom hierarchy to inside it, you will be replacing the entire dom of the parent. Real world example would be moving an entity in the level to a large prefab with 100s of entities. With this implementation, you will be replacing the dom of the level, which is the common parent.

The inverse of a move patch operation should just be another move patch operation right? It doesn't even have to be 2 replace operations if I understand the design correctly

Reply:
A move operation can potentially replace the contents at the destination. Per the JSON patch spec it's identical to a "remove" followed by an "add", which unfortunately makes it impossible to perfectly undo with one operation. If desired later on we can add a special case undo for move that performs a move / add by having this return something like a 2 entry max `fixed_vector` instead of a single patch.